DocumentsFromSnapshotMigration/src/test/java/org/opensearch/migrations/bulkload/CustomTransformationTest.java

```diff
@@ -0,0 +1,282 @@
+@Tag("isolatedTest")
+public class CustomTransformationTest extends SourceTestBase {
+
+    public static final String TARGET_DOCKER_HOSTNAME = "target";
+    public static final String SNAPSHOT_NAME = "test_snapshot";
+
+    @AllArgsConstructor
+    @Getter
+    private static class RunData {
+        Path tempDirSnapshot;
+        Path tempDirLucene;
+        SearchClusterContainer targetContainer;
+    }
+
+    @Test
+    public void testCustomTransformationProducesDesiredTargetClusterState() {
+        String nameTransformation = createIndexNameTransformation("geonames", "geonames_transformed");
+        var expectedSourceMap = new HashMap<String, Integer>();
+        expectedSourceMap.put("geonames", 1);
+        var expectedTargetMap = new HashMap<String, Integer>();
+        expectedTargetMap.put("geonames_transformed", 1);
+        // 2 Shards, for each shard, expect three status code 2 and one status code 0
+        int shards = 2;
+        int migrationProcessesPerShard = 4;
+        int continueExitCode = 2;
+        int finalExitCodePerShard = 0;
+        runTestProcessWithCheckpoint(continueExitCode, (migrationProcessesPerShard - 1) * shards,
+                finalExitCodePerShard, shards, expectedSourceMap, expectedTargetMap,
+            d -> runProcessAgainstTarget(d.tempDirSnapshot, d.tempDirLucene, d.targetContainer, nameTransformation
+            ));
+    }
+
+    @SneakyThrows
+    private void runTestProcessWithCheckpoint(int initialExitCode, int initialExitCodes,
+                                              int eventualExitCode, int eventualExitCodeCount,
+                                              Map<String, Integer> expectedSourceDocs,
+                                              Map<String, Integer> expectedTargetDocs,
+                                              Function<RunData, Integer> processRunner) {
+        final var testSnapshotContext = SnapshotTestContext.factory().noOtelTracking();
+
+        var tempDirSnapshot = Files.createTempDirectory("opensearchMigrationReindexFromSnapshot_test_snapshot");
+        var tempDirLucene = Files.createTempDirectory("opensearchMigrationReindexFromSnapshot_test_lucene");
+
+        try (
+            var esSourceContainer = new SearchClusterContainer(SearchClusterContainer.ES_V7_10_2)
+                    .withAccessToHost(true);
+            var network = Network.newNetwork();
+            var osTargetContainer = new SearchClusterContainer(SearchClusterContainer.OS_V2_14_0)
+                    .withAccessToHost(true)
+                    .withNetwork(network)
+                    .withNetworkAliases(TARGET_DOCKER_HOSTNAME);
+        ) {
+            CompletableFuture.allOf(
+                CompletableFuture.runAsync(esSourceContainer::start),
+                CompletableFuture.runAsync(osTargetContainer::start)
+            ).join();
+
+            var sourceClusterOperations = new ClusterOperations(esSourceContainer.getUrl());
+
+            var shards = 2;
```

Comment:
Non-blocking: why define this shard count again?  Maybe you can use a class-level member instead of redefining.

Reply:
This should be removed now, I've done a good amount of refactoring to remove extra logic for this test class that isn't needed